arroyo-worker/src/connectors/two_phase_committer.rs

```diff
@@ -0,0 +1,202 @@
+use std::{collections::HashMap, fmt::Debug, marker::PhantomData, time::SystemTime};
+
+use anyhow::Result;
+use arroyo_macro::{process_fn, StreamNode};
+use arroyo_rpc::{
+    grpc::{TableDeleteBehavior, TableDescriptor, TableType, TableWriteBehavior},
+    CheckpointEvent, ControlMessage,
+};
+use arroyo_state::tables::GlobalKeyedState;
+use async_trait::async_trait;
+use serde::{Deserialize, Serialize};
+use tracing::{info, warn};
+use typify::import_types;
+
+import_types!(schema = "../connector-schemas/parquet/table.json");
+
+use crate::engine::Context;
+use arroyo_types::*;
+
+#[derive(StreamNode)]
+pub struct TwoPhaseCommitterOperator<K: Key, T: Data + Sync, TPC: TwoPhaseCommitter<K, T>> {
+    committer: TPC,
+    pre_commits: Vec<TPC::PreCommit>,
+    phantom: PhantomData<(K, T)>,
+}
+
+/// A trait representing a two-phase committer for a stream processing system.
+///
+/// This trait defines the interface for a two-phase committer, which is responsible for committing
+/// records to a persistent store in a fault-tolerant manner. The two-phase commit protocol is used
+/// to ensure that all records are either committed or rolled back in the event of a failure.
+///
+/// The trait defines methods for initializing the committer, inserting records, committing the
+/// records, and performing a checkpoint. Implementations of this trait must be `Send` and `'static`.
+///
+/// The trait is generic over two types: `K`, which represents the key type of the records being
+/// committed, and `T`, which represents the data type of the records being committed. The trait
+/// also defines two associated types: `DataRecovery`, which represents the type of data that can
+/// be recovered in the event of a failure, and `PreCommit`, which represents the type of data that
+/// is pre-committed before the final commit.
+///
+///
+#[async_trait]
+pub trait TwoPhaseCommitter<K: Key, T: Data + Sync>: Send + 'static {
+    type DataRecovery: Data;
+    type PreCommit: Data;
+
+    fn name(&self) -> String;
+    async fn init(
+        &mut self,
+        task_info: &TaskInfo,
+        data_recovery: Vec<Self::DataRecovery>,
+    ) -> Result<()>;
+    async fn insert_record(&mut self, record: &Record<K, T>) -> Result<()>;
+    // TODO: figure out how to have the relevant vectors be of pointers across async boundaries.
```

Comment:
they can be pointers across async bounderies so long as they are `&mut` (because `&T` is send only if `T` is sync, but `&mut T` is send if `T` is send).